torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,401 @@
+                // invoked with error, and shouldn't read here.
+                TORCH_INTERNAL_ASSERT(!readError_, "Shouldn't in error state");
+                auto it = pendingResponseMessage_.find(messageId);
+                TORCH_INTERNAL_ASSERT(
+                    it != pendingResponseMessage_.end(),
+                    "message ID ",
+                    messageId,
+                    " is not recognized");
+                futureResponseMessage = std::move(it->second);
+                pendingResponseMessage_.erase(it);
+              }
+
+              if (error) {
+                LOG(WARNING) << "Read response error: " << error.what();
+                futureResponseMessage->setError(error.what());
+                return;
+              }
+
+              if (responseMessage.type() == MessageType::EXCEPTION) {
+                futureResponseMessage->setError(std::string(
+                    responseMessage.payload().begin(),
+                    responseMessage.payload().end()));
+              } else {
+                futureResponseMessage->markCompleted(
+                    std::move(responseMessage));
+              }
+            });
+      });
+
+  return futureResponseMessage;
+}
+
+// TODO: Remove sync()
+void TensorPipeAgent::sync() {}
+
+// TODO: Remove join()
+void TensorPipeAgent::join() {
+  shutdownImpl();
+}
+
+void TensorPipeAgent::shutdownImpl() {
+  isRunning_ = false;
+  // TODO: context_->join() is not absolutely ready yet.
+  // NOTE: context_->join() will wait for available RPC message to be
+  //       read or written, and wait for the remaining unavailable ones
+  //       to be called with error by invoking callbacks.
+}
+
+const WorkerInfo& TensorPipeAgent::getWorkerInfo(
+    const std::string& workerName) const {
+  const auto& it = workerNameToInfo_.find(workerName);
+  if (it != workerNameToInfo_.end()) {
+    return it->second;
+  }
+  TORCH_CHECK(false, "Unknown destination worker ", workerName);
+}
+
+const WorkerInfo& TensorPipeAgent::getWorkerInfo(worker_id_t workerId) const {
+  const auto& it = workerIdToInfo_.find(workerId);
+  if (it != workerIdToInfo_.end()) {
+    return it->second;
+  }
+  TORCH_CHECK(false, "Unknown destination worker ", workerId);
+}
+
+std::vector<WorkerInfo> TensorPipeAgent::getWorkerInfos() const {
+  std::vector<WorkerInfo> workerInfos;
+  for (auto& item : workerNameToInfo_) {
+    workerInfos.emplace_back(item.second);
+  }
+  return workerInfos;
+}
+
+const std::string& TensorPipeAgent::findWorkerURL(
+    const WorkerInfo& worker) const {
+  const auto it = workerNameToURL_.find(worker.name_);
+  TORCH_CHECK(
+      it != workerNameToURL_.end(), "Unknown worker name: ", worker.name_);
+  return it->second;
+}
+
+std::string TensorPipeAgent::createUniqueShmAddr() {
+  return c10::str(
+      "shm://tensorpipe_rpc_agent_",
+      std::this_thread::get_id(),
+      "_",
+      nextShmAddressID_++);
```

Comment:
It's possible that we generate the same shm addr for two different processes on the same host since the first time `nextShmAddressID_` would be 0 for both and since thread ids can be recycled, `std::this_thread::get_id()` can potentially return the same id. Also, I'm not sure if `std::this_thread::get_id()` is guaranteed to be unique across different processes.

I think we should probably just refactor and reuse: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_context.h#L61 and get rid of `std::this_thread::get_id()` and `nextShmAddressID_` here.

Reply:
Seems it's using getWorkerId() and nextLocalId to create globally unique identifier. If we have multiple training groups on a single host, worker_id could conflict as well? LocalId also doesn't provide uniqueness across multiple processes. Maybe we can use process ID + thread ID + local ID? As shm is just for linux for now, it should be fine to have some platform specific code?

Reply:
Worker id is supposed to be unique across a training group. And I think different training groups should use different store instances or if they use the same store instance, they should use a different prefix in the store to avoid conflicts.